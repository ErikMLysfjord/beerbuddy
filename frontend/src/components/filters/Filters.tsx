import { Button, Checkbox, Collapse, Divider, Tooltip } from "antd";
import { Slider, SliderProps, alpha, styled } from "@mui/material";
import { useContext, useId, useRef } from "react";
import { FilterContext } from "../../context/FilterContext";
import {
  QuestionCircleOutlined,
  RollbackOutlined,
  RightOutlined,
} from "@ant-design/icons";
import { default as CSSstyles } from "./Filters.module.css";

// Styles of beer that can be filtered on
const beerStyles = [
  "American IPA",
  "American Pale Ale (APA)",
  "American Amber / Red Ale",
  "American Blonde Ale",
  "American Double / Imperial IPA",
  "American Pale Wheat Ale",
  "American Brown Ale",
  "American Porter",
  "Saison / Farmhouse Ale",
  "Witbier",
  "Fruit / Vegetable Beer",
  "KÃ¶lsch",
  "Hefeweizen",
  "American Pale Lager",
  "American Stout",
  "Other",
];

// Styled slider component from MUI
const StyledSlider = styled(Slider)<SliderProps>(() => ({
  color: "#FFCC48",
  "& .MuiSlider-thumb": {
    "&:hover, &.Mui-focusVisible": {
      boxShadow: `0px 0px 0px 8px ${alpha("#FFCC48", 0.16)}`,
    },
    "&.Mui-active": {
      boxShadow: `0px 0px 0px 14px ${alpha("#FFCC48", 0.16)}`,
    },
  },
}));

interface FiltersProps {
  fetchMore: (reset?: boolean) => Promise<void>;
  apply?: () => void;
}
/**
 * Filters component that contains all filters.
 * @param fetchMore - function that is called when the apply filters button is clicked
 * @param apply - function that is called to hide the filters when the apply filters button is clicked
 * @returns a Filters component
 */
const Filters = ({ fetchMore, apply }: FiltersProps) => {
  const {
    IBU,
    setIBU,
    ABV,
    setABV,
    setStyles,
    styles,
    searchString,
    setSearchString,
  } = useContext(FilterContext);
  const fetchMoreRef = useRef(fetchMore);
  fetchMoreRef.current = fetchMore;
  const headingID = useId();

  /**
   * Function for resetting filters and fetching more beers.
   */
  const resetFilters = () => {
    /* Only reset filters and fetch beers if there are active filters */
    if (
      beerStyles.length > 0 ||
      ABV[0] !== 0 ||
      ABV[1] !== 40 ||
      IBU[0] !== 0 ||
      IBU[1] !== 140 ||
      searchString !== ""
    ) {
      setABV([0, 40]);
      setIBU([0, 140]);
      setStyles([]);
      setSearchString("");
      fetchMoreRef.current(true);
    }
  };

  return (
    <section aria-labelledby={headingID}>
      <Divider />
      <h2 id={headingID}>Filters</h2>

      <Collapse
        aria-label="Beer styles"
        expandIcon={({ isActive }) => (
          <RightOutlined rotate={isActive ? 90 : 0} aria-hidden />
        )}
        className={CSSstyles.collapse}
        items={[
          {
            key: "styles-checkbox-key",
            label: (
              <span className={CSSstyles.collapseSpan}>
                Beer styles
                <Tooltip title={"Filter on certain styles of beer."}>
                  <QuestionCircleOutlined aria-hidden />
                </Tooltip>
              </span>
            ),
            children: (
              <Checkbox.Group
                options={beerStyles}
                name="styles"
                onChange={(checkedValues) =>
                  setStyles(checkedValues.map((value) => value.toString()))
                }
                value={styles}
              />
            ),
          },
        ]}
      />
      <Collapse
        expandIcon={({ isActive }) => (
          <RightOutlined rotate={isActive ? 90 : 0} aria-hidden />
        )}
        className={CSSstyles.collapse}
        items={[
          {
            key: "IBU-range-key",
            label: (
              <span className={CSSstyles.collapseSpan}>
                IBU
                <Tooltip title={"Filter on the bitterness of the beer."}>
                  <QuestionCircleOutlined aria-hidden />
                </Tooltip>
              </span>
            ),
            children: (
              <StyledSlider
                getAriaLabel={() => "IBU range"}
                value={IBU}
                onChange={(e: Event, newValue: number | number[]) => {
                  e.preventDefault();
                  setIBU(newValue as number[]);
                }}
                max={140}
                valueLabelDisplay="auto"
                getAriaValueText={() =>
                  `Min value is ${IBU[0]}, max is ${IBU[1]}`
                }
              />
            ),
          },
        ]}
      />
      <Collapse
        className={CSSstyles.collapse}
        expandIcon={({ isActive }) => (
          <RightOutlined rotate={isActive ? 90 : 0} aria-hidden />
        )}
        items={[
          {
            key: "ABV-range-key",
            label: (
              <span className={CSSstyles.collapseSpan}>
                ABV
                <Tooltip
                  title={"Filter on the alcohol percentage of the beer."}
                >
                  <QuestionCircleOutlined aria-hidden />
                </Tooltip>
              </span>
            ),
            children: (
              <StyledSlider
                getAriaLabel={() => "ABV range"}
                value={ABV}
                onChange={(e: Event, newValue: number | number[]) => {
                  e.preventDefault();
                  setABV(newValue as number[]);
                }}
                max={40}
                valueLabelDisplay="auto"
                getAriaValueText={() =>
                  `Min value is ${ABV[0]}, max is ${ABV[1]}`
                }
              />
            ),
          },
        ]}
      />

      <Button
        type="primary"
        className={CSSstyles.buttons}
        onClick={(e) => {
          e.preventDefault();
          if (apply) apply();
          fetchMore(true);
        }}
      >
        Apply Filters
      </Button>
      <Button
        type="default"
        className={CSSstyles.buttons}
        onClick={resetFilters}
        icon={<RollbackOutlined />}
      >
        Reset Filters
      </Button>
    </section>
  );
};
export default Filters;
