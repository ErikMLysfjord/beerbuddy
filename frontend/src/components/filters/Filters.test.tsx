import { it, expect, describe, vi, vitest } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Filters from "./Filters";
import { act } from "react-dom/test-utils";
import { axe } from "jest-axe";
import { useState as useStateMock } from "react";

vi.mock("react", () => ({
  ...vitest.importActual("react"),
  useState: vi.fn(),
}));

describe("Filters", () => {
  const setState = vi.fn();

  beforeEach(() => {
    (useStateMock as jest.Mock).mockImplementation((init) => [init, setState]);
  });

  it("is accessible", async () => {
    const { container } = render(<Filters />);
    expect(await axe(container)).toHaveNoViolations();
  });

  it("renders correctly", () => {
    const { container } = render(<Filters />);
    expect(container).toMatchSnapshot();
  });

  it("renders a Filter component for each filter", () => {
    const { container } = render(<Filters />);
    expect(container.querySelectorAll(".ant-divider").length).toBe(3);
  });

  it("renders a checkbox group for the beer styles", () => {
    render(<Filters />);
    expect(screen.getAllByRole("checkbox").length).toBe(16);

    const firstCheckbox = screen.getByLabelText("American IPA");
    expect(firstCheckbox).toBeInTheDocument();
    expect(firstCheckbox).not.toBeChecked();

    act(() => {
      firstCheckbox.dispatchEvent(new MouseEvent("click", { bubbles: true }));
    });

    expect(firstCheckbox).toBeChecked();
  });

  it("renders a slider for the IBU", () => {
    render(<Filters />);
    expect(screen.queryAllByRole("slider")[0]).toBeInTheDocument();
  });

  it("renders a filter for style, ibu and alcohol", () => {
    render(<Filters />);
    expect(screen.getByText("Style")).toBeInTheDocument();
    expect(screen.getByText("Style")).toBeInstanceOf(HTMLHeadingElement);

    expect(screen.getByText("IBU")).toBeInTheDocument();
    expect(screen.getByText("IBU")).toBeInstanceOf(HTMLHeadingElement);

    expect(screen.getByText("Alcohol")).toBeInTheDocument();
    expect(screen.getByText("Alcohol")).toBeInstanceOf(HTMLHeadingElement);
  });

  it("changes value of IBU slider", () => {
    render(<Filters />);
    const minSlider = screen.getAllByRole("slider")[0];
    const maxSlider = screen.getAllByRole("slider")[1];
    act(() => {
      fireEvent.change(minSlider, { target: { value: 25 } });
    });
    expect(setState).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledWith([25, 140]);

    setState.mockReset();
    act(() => {
      fireEvent.change(maxSlider, { target: { value: 30 } });
    });

    expect(setState).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledWith([0, 30]);
    setState.mockReset();
  });

  it("changes value of alcohol slider", () => {
    render(<Filters />);
    const minSlider = screen.getAllByRole("slider")[2];
    const maxSlider = screen.getAllByRole("slider")[3];
    act(() => {
      fireEvent.change(minSlider, { target: { value: 10 } });
    });
    expect(setState).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledWith([10, 40]);

    setState.mockReset();
    act(() => {
      fireEvent.change(maxSlider, { target: { value: 30 } });
    });

    expect(setState).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledWith([0, 30]);
  });
});
