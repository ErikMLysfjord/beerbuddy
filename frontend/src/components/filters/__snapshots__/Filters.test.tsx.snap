// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filters > renders correctly 1`] = `
<div>
  <div
    class="ant-divider css-dev-only-do-not-override-pr0fja ant-divider-horizontal"
    role="separator"
  />
  <div
    class="_headingWrapper_3ea492"
  >
    <h2>
      Style
    </h2>
    <span
      aria-label="question-circle"
      class="anticon anticon-question-circle"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="question-circle"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
        <path
          d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
        />
      </svg>
    </span>
  </div>
  <div
    class="ant-checkbox-group css-dev-only-do-not-override-pr0fja"
  >
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="IPA"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        IPA
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Stout"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Stout
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Lager"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Lager
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Pilsner"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Pilsner
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Pale Ale"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Pale Ale
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Wheat"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Wheat
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Sour"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Sour
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Porter"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Porter
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Belgian"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Belgian
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Brown Ale"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Brown Ale
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Red Ale"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Red Ale
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Barleywine"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Barleywine
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Saison"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Saison
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Bock"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Bock
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="Amber Ale"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        Amber Ale
      </span>
    </label>
    <label
      class="ant-checkbox-wrapper ant-checkbox-group-item css-dev-only-do-not-override-pr0fja"
    >
      <span
        class="ant-checkbox ant-wave-target css-dev-only-do-not-override-pr0fja"
      >
        <input
          class="ant-checkbox-input"
          type="checkbox"
          value="APA"
        />
        <span
          class="ant-checkbox-inner"
        />
      </span>
      <span>
        APA
      </span>
    </label>
  </div>
  <div
    class="ant-divider css-dev-only-do-not-override-pr0fja ant-divider-horizontal"
    role="separator"
  />
  <div
    class="_headingWrapper_3ea492"
  >
    <h2>
      IBU
    </h2>
    <span
      aria-label="question-circle"
      class="anticon anticon-question-circle"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="question-circle"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
        <path
          d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
        />
      </svg>
    </span>
  </div>
  <div
    class="ant-slider css-dev-only-do-not-override-pr0fja ant-slider-horizontal"
  >
    <div
      class="ant-slider-rail"
    />
    <div
      class="ant-slider-track ant-slider-track-1"
      style="left: 0%; width: 24%;"
    />
    <div
      class="ant-slider-step"
    />
    <div
      aria-disabled="false"
      aria-orientation="horizontal"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="ant-slider-handle ant-slider-handle-1"
      role="slider"
      style="left: 0%; transform: translateX(-50%);"
      tabindex="0"
    />
    <div
      aria-disabled="false"
      aria-orientation="horizontal"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="24"
      class="ant-slider-handle ant-slider-handle-2"
      role="slider"
      style="left: 24%; transform: translateX(-50%);"
      tabindex="0"
    />
  </div>
  <div
    class="ant-divider css-dev-only-do-not-override-pr0fja ant-divider-horizontal"
    role="separator"
  />
  <div
    class="_headingWrapper_3ea492"
  >
    <h2>
      Alcohol
    </h2>
    <span
      aria-label="question-circle"
      class="anticon anticon-question-circle"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="question-circle"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
        <path
          d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
        />
      </svg>
    </span>
  </div>
  <div
    class="ant-slider css-dev-only-do-not-override-pr0fja ant-slider-horizontal"
  >
    <div
      class="ant-slider-rail"
    />
    <div
      class="ant-slider-track ant-slider-track-1"
      style="left: 16%; width: 16%;"
    />
    <div
      class="ant-slider-step"
    />
    <div
      aria-disabled="false"
      aria-orientation="horizontal"
      aria-valuemax="25"
      aria-valuemin="0"
      aria-valuenow="4"
      class="ant-slider-handle ant-slider-handle-1"
      role="slider"
      style="left: 16%; transform: translateX(-50%);"
      tabindex="0"
    />
    <div
      aria-disabled="false"
      aria-orientation="horizontal"
      aria-valuemax="25"
      aria-valuemin="0"
      aria-valuenow="8"
      class="ant-slider-handle ant-slider-handle-2"
      role="slider"
      style="left: 32%; transform: translateX(-50%);"
      tabindex="0"
    />
  </div>
</div>
`;
